web/app/dao/SearchDAO.java

```diff
@@ -183,34 +164,94 @@
-		List<String> names = new ArrayList<>();
-		if (limit <= 0) {
-			names.addAll(getJdbcTemplate().queryForList(GET_FLOW_AUTO_COMPLETE_LIST_NO_LIMIT, String.class, pattern));
-			names.addAll(getJdbcTemplate().queryForList(GET_JOB_AUTO_COMPLETE_LIST_NO_LIMIT, String.class, pattern));
-		} else {
-			names.addAll(getJdbcTemplate().queryForList(GET_FLOW_AUTO_COMPLETE_LIST, String.class, pattern, limit / 2));
-			names.addAll(getJdbcTemplate().queryForList(GET_JOB_AUTO_COMPLETE_LIST, String.class, pattern, limit / 2));
-		}
-		return names;
+		String elasticSearchTypeURLKey = "elasticsearch.flow.url";
+		String fieldName = "flow_name_suggest";
+		return getAutoCompleteListbyES(elasticSearchTypeURLKey,fieldName,input,limit);
 	}
 
-	public static List<String> getAutoCompleteListMetric(String input, int limit) {
-		String pattern = isNotBlank(input) ? input + "%" : "%";
-		if (limit <= 0) {
-			return getJdbcTemplate().queryForList(GET_METRIC_AUTO_COMPLETE_LIST_NO_LIMIT, String.class, pattern);
+	public static List<String> getAutoCompleteListbyES(String elasticSearchTypeURLKey, String fieldName, String input, int limit) {
+		// use elastic search completion suggester, ES will validate the input and limit
+		List<String> completionSuggestionList = new ArrayList<String>();
+		JsonNode responseNode = null;
+		ObjectNode keywordNode = null;
+
+		try
+		{
+			keywordNode = utils.Search.generateElasticSearchCompletionSuggesterQuery(fieldName, input,limit);
+		}
+		catch(Exception e)
+		{
+			Logger.error("Elastic search completion suggester error. Error message :" + e.getMessage());
 		}
-		return getJdbcTemplate().queryForList(GET_METRIC_AUTO_COMPLETE_LIST, String.class, pattern, limit);
+
+		Logger.info("The completion suggester query sent to Elastic Search is: " + keywordNode.toString());
+
+		Promise<WSResponse> responsePromise = WS.url(Play.application().configuration().getString(
+				elasticSearchTypeURLKey)).post(keywordNode);
+		responseNode = responsePromise.get(1000).asJson();
+
+		if (responseNode == null || !responseNode.isContainerNode())
+			return completionSuggestionList;
+
+		JsonNode suggestNode = responseNode.get("suggest");
+		if (suggestNode == null || !suggestNode.has("wh-suggest"))
+		{
+			Logger.error("Elastic search completion suggester response does not contain suggest node");
+			return completionSuggestionList;
+		}
+
+		Logger.info("Response suggestNode is " + suggestNode.toString());
+
+		JsonNode whSuggestNode = suggestNode.get("wh-suggest");
+		if (whSuggestNode == null || !whSuggestNode.isArray())
+		{
+			Logger.error("Elastic search completion suggester response does not contain wh-suggest node");
+			return completionSuggestionList;
+		}
+
+		Iterator<JsonNode> arrayIterator = whSuggestNode.elements();
+		if (arrayIterator == null) return completionSuggestionList;
+
+		while (arrayIterator.hasNext())
+		{
+			JsonNode node = arrayIterator.next();
+			if (node.isContainerNode() && node.has("options"))
```

Comment:
Let's get rid of the nested if statements with early terminations, i.e. something like if (!node.isContainerNode() || !node.has("options")) continue;  to improve readability.